mayastor/src/subsys/mbus/registration.rs

```diff
@@ -0,0 +1,201 @@
+use once_cell::sync::OnceCell;
+use serde::{Deserialize, Serialize};
+use snafu::Snafu;
+use std::{env, time::Duration};
+
+use crate::subsys::MessageBus;
+
+/// Mayastor sends registration messages in this interval (kind of heart-beat)
+const HB_INTERVAL: Duration = Duration::from_secs(2);
+
+/// Errors for pool operations.
+///
+/// Note: The types here that would be normally used as source for snafu errors
+/// do not implement Error trait required by Snafu. So they are renamed to
+/// "cause" attribute and we use .map_err() instead of .context() when creating
+/// them.
+#[derive(Debug, Snafu)]
+pub enum Error {
+    #[snafu(display(
+        "Failed to connect to the MessageBus server {}: {:?}",
+        server,
+        cause
+    ))]
+    ConnectFailed {
+        cause: std::io::Error,
+        server: String,
+    },
+    #[snafu(display(
+        "Cannot issue requests if message bus hasn't been started"
+    ))]
+    NotStarted {},
+    #[snafu(display("Failed to queue register request: {:?}", cause))]
+    QueueRegister { cause: std::io::Error },
+    #[snafu(display("Failed to queue deregister request: {:?}", cause))]
+    QueueDeregister { cause: std::io::Error },
+}
+
+/// Register message payload
+#[derive(Serialize, Deserialize, Debug)]
+struct RegisterArgs {
+    id: String,
+    #[serde(rename = "grpcEndpoint")]
+    grpc_endpoint: String,
+}
+
+/// Deregister message payload
+#[derive(Serialize, Deserialize, Debug)]
+struct DeregisterArgs {
+    id: String,
+}
+
+#[derive(Clone)]
+struct Configuration {
+    /// Name of the node that mayastor is running on
+    node: String,
+    /// gRPC endpoint of the server provided by mayastor
+    grpc_endpoint: String,
+    /// heartbeat interval (how often the register message is sent)
+    hb_interval: Duration,
+}
+
+#[derive(Clone)]
+pub(super) struct Registration {
+    /// Configuration of the registration
+    config: Configuration,
+    /// Receive channel for messages and termination
+    rcv_chan: smol::channel::Receiver<()>,
+    /// Termination channel
+    fini_chan: smol::channel::Sender<()>,
+}
+
+static MESSAGE_BUS_REG: OnceCell<Registration> = OnceCell::new();
+impl Registration {
+    pub(super) fn init(node: &str, grpc_endpoint: &str) {
+        MESSAGE_BUS_REG.get_or_init(|| Registration::new(node, grpc_endpoint));
+
+        // spawn a runner thread responsible for registering and
+        // deregistering the mayastor instance on shutdown
+        std::thread::spawn(|| {
+            smol::block_on(async {
+                Self::get().clone().run().await;
+            });
+        });
+    }
+
+    fn new(node: &str, grpc_endpoint: &str) -> Registration {
+        let (msg_sender, msg_receiver) = smol::channel::unbounded::<()>();
+        let config = Configuration {
+            node: node.to_owned(),
+            grpc_endpoint: grpc_endpoint.to_owned(),
+            hb_interval: match env::var("MAYASTOR_HB_INTERVAL")
+                .map(|v| v.parse::<u64>())
```

Comment:
map_or_else() or unwrap_or_default might be useful here?

Reply:
Had trouble making it "neat" that way because env::var returns a result and parse also returns a result so ended up having to do two unwraps, eg: